content/2021-06-11-make-your-life-easier.md

```diff
@@ -0,0 +1,156 @@
++++
+title = "Make your life easier"
+template = "page.html"
+date = 2021-06-11T22:00:00Z
+
+[taxonomies]
+tags = ["Tech"]
++++
+
+If you write software for long enough you begin to develop instincts about the
+relative difficulty of different problems and some of those solidify into
+rules of thumb. Like all such things they are not really _rules_ but patterns
+of thinking that you can choose to deviate from when you have good reason to
+do so. The following are some that I have come across recently. I can't claim
+that any of them are original ideas, but they are some of the things that
+guide my work on a daily basis.
+
+# Choose simpler graph data structures
+
+Many problems are graph problems of one sort or another and there's a
+hierarchy of increasing complexity of graph types:
+
+- list
+- tree
+- acyclic graph
+- cyclic graph
+
+All of those are technically graphs, but the amount of work you need to do to
+operate on them gets noticeably harder as you work your way down the list. If
+you can constrain your requirements to allow you to use structures nearer the
+top of that list you will have fewer implementation bugs.
+
+Sometimes you can't avoid a cyclic graph, but by remembering that it's harder
+than the others you should stop and think "do I need more tests for this?"
+
+# Default to immutable data
+
+Cache invalidation is hard. If your data is immutable you rarely have to solve
+that problem. Immutable data is also easier to archive and/or store in cheap
+blob stores like S3 or in static files on a filesystem.
+
+It can be very hard to make your data truly immutable. For example, the [GDPR
+"right to be forgotten"](https://gdpr-info.eu/art-17-gdpr/) means that
+anything containing [personal data](https://gdpr-info.eu/art-4-gdpr/) must be
+deletable. So just like with graph data structures there's a kind of hierarchy
+of mutability of data structures:
+
+- Truly immutable. The data is created and is never destroyed.
+- Can be atomicly created or destroyed but never modified.
+- Can have portions appended or deleted, but previous data is never modified.
+- Some portions of the data structure are immutable, others can be deleted or
+  mutated in place.
+- All parts of the data structure are mutable.
+
+Stay as far up that list as you can and you will avoid many implementation
```

Comment:
What is a practical example of this? I feel like, with Clojure, this makes a ton of sense, but it may not be as clear to someone working in, say, Python.

Reply:
I was thinking more about data at rest than anything language-specific. I'll try to add some clarifying language.